runtime_lib/aircpu/channel.cpp

```diff
@@ -0,0 +1,239 @@
+// Copyright (C) 2023, Advanced Micro Devices, Inc. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#include "air_channel.h"
+#include "air_tensor.h"
+
+#include <iostream>
+
+#define VERBOSE 0
+
+template <typename T, int R>
+static void _air_channel_put(tensor_t<uint64_t, 0> *channel,
+                             tensor_t<T, R> *src, size_t *_offset,
+                             size_t *_size, size_t *_stride) {
+  size_t offset[4] = {0, 0, 0, 0};
+  size_t size[4] = {1, 1, 1, 1};
+  size_t stride[4] = {1, 1, 1, 1};
+  for (int i = 0; i < R; i++) {
+    offset[i] = _offset[i];
+    size[i] = _size[i];
+    stride[i] = _stride[i];
+  }
+
+  // channel->data is an array of pointers to channel_t objects
+  // if the channel is valid, channel->data[0] should be a valid memory address
+  // otherwise, allocate a new channel
+  if (channel->data[0] == 0) {
+    channel_t<T> *new_channel = new channel_t<T>(size);
+    channel->data[0] = (uint64_t)new_channel;
+  }
+
+  channel_t<T> *chan = (channel_t<T> *)channel->data[0];
+
+  // wait until the channel is empty
+  std::unique_lock<std::mutex> lock(chan->mtx);
+  chan->cv.wait(lock, [&chan] { return !chan->_is_full; });
```

Comment:
Synchronization between put and get is now implemented with conditional variables